// @cloudscape-design/components/test-utils/selectors/pagination/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaginationButtonWrapper = void 0;
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
const selectors_1 = require("@cloudscape-design/test-utils-core/selectors");
const button_1 = require("../button");
const input_1 = require("../input");
const popover_1 = require("../popover");
const styles_selectors_js_1 = require("../../../pagination/styles.selectors.js");
class PaginationButtonWrapper extends selectors_1.ComponentWrapper {
}
exports.PaginationButtonWrapper = PaginationButtonWrapper;
class PaginationWrapper extends selectors_1.ComponentWrapper {
    findCurrentPage() {
        return this.findComponent(`.${styles_selectors_js_1.default['button-current']}`, PaginationButtonWrapper);
    }
    findPageNumbers() {
        return this.findAllByClassName(styles_selectors_js_1.default['page-number']).map(wrapper => new PaginationButtonWrapper(wrapper.getElement()));
    }
    /**
     * Returns a page number for a given index.
     *
     * @param index 1-based index of the page number to return.
     */
    findPageNumberByIndex(index) {
        // we need to skip the "previous page" button
        const pageIndex = index + 1;
        return this.findComponent(`li:nth-child(${pageIndex}) .${styles_selectors_js_1.default.button}`, PaginationButtonWrapper);
    }
    findPreviousPageButton() {
        return this.findComponent(`li:first-child .${styles_selectors_js_1.default.button}`, PaginationButtonWrapper);
    }
    findNextPageButton() {
        return this.findComponent(`li:last-child .${styles_selectors_js_1.default.button}`, PaginationButtonWrapper);
    }
    /**
     * Returns the jump to page input field.
     */
    findJumpToPageInput() {
        return this.findComponent(`.${styles_selectors_js_1.default['jump-to-page-input']}`, input_1.default);
    }
    /**
     * Returns the jump to page submit button.
     */
    findJumpToPageButton() {
        const jumpToPageContainer = this.findByClassName(styles_selectors_js_1.default['jump-to-page']);
        return jumpToPageContainer ? jumpToPageContainer.findComponent('button', button_1.default) : null;
    }
    /**
     * Returns the error popover for jump to page.
     */
    findJumpToPagePopover() {
        return this.findComponent(`.${popover_1.default.rootSelector}`, popover_1.default);
    }
}
PaginationWrapper.rootSelector = styles_selectors_js_1.default.root;
exports.default = PaginationWrapper;
